### PR TITLE
Lwg 2024 02 29 updates

### PR DESCRIPTION
**Pull request by crtrott**

### CODE REVIEW

**Review by dalg24 — APPROVED** [review]

**Comment by mhoemmen** [comment]
Thanks @crtrott !  There was one more change request that's not in this PR.  It's OK to merge this one and do that later if you like, or we can just put it in this PR -- it's up to you.

> [mdspan.submdspan.mapping]
>
> Add a drafting Note to explain that 4.2 is a drive-by fix, an intentional deviation from the current Draft that LWG reviewed and approved.

**Review by mhoemmen — APPROVED** [review]

File: layout_padded/layout_padded.md [hunk 1]
```diff
@@ -2529,17 +2531,20 @@ is `true`.  Otherwise, `false`.
 
 [1]{.pnum} *Returns:*
 
-   * [1.1]{.pnum} `submdspan_mapping_result{*this, 0}`, if `Extents::rank()==0` is `true`;
+   * [1.1]{.pnum} `submdspan_mapping_result{*this, 0}`, if _`rank_`_`==0` is `true`;
+   
+   * [1.2]{.pnum} otherwise, `submdspan_mapping_result{layout_left::mapping(sub_ext), offset}`, if `SubExtents::rank()==0` is `true`;
```
Comment (mhoemmen) [inline layout_padded/layout_padded.md:2536]: Note to self to check that all the layout mappings' _`submdspan-mapping-impl`_ implementations have this optimization.

File: layout_padded/layout_padded.md [hunk 2]
```diff
@@ -2651,18 +2675,35 @@ and where `s_dynamic` is the product of all values `extent(k)` for `k` in the ra
...
 
-   * [1.1]{.pnum} `submdspan_mapping_result{*this, 0}`, if `Extents::rank() == 0` is `true`;
+   * [1.1]{.pnum} `submdspan_mapping_result{*this, 0}`, if _`rank_`_` == 0` is `true`;
+
+   * [1.2]{.pnum} otherwise, `submdspan_mapping_result{layout_right::mapping(sub_ext), offset}`, if
+
+      * _`rank_`_`==1` is `true`; or
 
-   * [1.2]{.pnum} otherwise, `submdspan_mapping_result{layout_right_padded<padding_value>::mapping(sub_ext), offset}`, if `Extents::rank() == 1` is `true`;
+      * `SubExtents::rank()==0` is `true`;
```
Comment (mhoemmen) [inline layout_padded/layout_padded.md:2684]: Should we consider using this two-bullet-point format for all the analogous cases above?  I know LWG doesn't really like two bullet points, but we should at least have the same number of numbered paragraphs (one, in this case: 1.2) for all the cases.  I'll suggest appropriate changes above.

Comment (mhoemmen) [reply layout_padded/layout_padded.md:2684]: Actually, never mind; I see why we might like to do this differently for the other layouts.

File: layout_padded/layout_padded.md [hunk 3]
```diff
@@ -2651,18 +2675,35 @@ and where `s_dynamic` is the product of all values `extent(k)` for `k` in the ra
...
+   
+   * [1.3]{.pnum} otherwise, `submdspan_mapping_result{layout_right::mapping(sub_ext), offset}`, if
 
-   * [1.3]{.pnum} otherwise, `submdspan_mapping_result{layout_right_padded<padding_value>::mapping(sub_ext, s_dynamic), offset}` if there exists an integral value `u` such that
+      * `SubExtents::rank()==1` is `true`; and
 
-      * for `k` equal to `Extents::rank() - 1`, $S_k$ models _`index-pair-like`_`<index_type>` or `is_convertible_v<`$S_k$`, full_extent_t>` is `true`; and
+      * for `k` equal to _`rank_`_` - 1`, $S_k$ models _`index-pair-like`_`<index_type>` or `is_convertible_v<` $S_k$ `, full_extent_t>` is `true`; and
```
Comment (mhoemmen) [inline layout_padded/layout_padded.md:2690]: Would you consider removing the dangling "and"?

```suggestion
      * for `k` equal to _`rank_`_` - 1`, $S_k$ models _`index-pair-like`_`<index_type>` or `is_convertible_v<` $S_k$ `, full_extent_t>` is `true`.
```

File: layout_padded/layout_padded.md [hunk 4]
```diff
@@ -2651,18 +2675,35 @@ and where `s_dynamic` is the product of all values `extent(k)` for `k` in the ra
...
+
+   * [1.4]{.pnum} otherwise, `submdspan_mapping_result{layout_right_padded<S_static>::template mapping(sub_ext, s_dynamic), offset}` 
+if for a value `u` for which _`rank_`_` - u - 2` is the largest value `k` smaller than _`rank_`_` - 1` for which $S_k$ models _`index-pair-like`_`<index_type>` or `is_convertible_v<`$S_k$`, full_extent_t>` is `true`, the following conditions are met:
+
+      * for `k` equal to _`rank_`_` - 1`, $S_k$ models _`index-pair-like`_`<index_type>` or `is_convertible_v<` $S_k$ `, full_extent_t>` is `true`; and
```
[the remark this answers is not in the document]
Comment (mhoemmen) [reply layout_padded/layout_padded.md:2695]: I think perhaps the better approach would be to use a different letter than `k` on line 2693.  That would free up `k` for these subbullet points to be a "local index variable."

File: layout_padded/layout_padded.md (hunk 1 above)
Comment (mhoemmen) [reply layout_padded/layout_padded.md:2536]: They do : - )

**Review by mhoemmen — APPROVED** [review]

File: layout_padded/layout_padded.md [hunk 5]
```diff
@@ -2121,7 +2124,7 @@ private:
 
 [5]{.pnum} *Mandates:*
 
-  * [5.1]{.pnum} If `Extents​::​rank_dynamic() == 0` is `true`, then the size of the multidimensional index space `Extents()` is representable as a value of type typename `Extents​::​index_type`.
+  * [5.1]{.pnum} If `rank_dynamic() == 0` is `true`, then the size of the multidimensional index space `Extents()` is representable as a value of type `index_type`.
```
Comment (mhoemmen) [inline layout_padded/layout_padded.md:2127]: The other one says `Extents::index_type`.

File: layout_padded/layout_padded.md [hunk 6]
```diff
@@ -1645,7 +1648,7 @@ private:
 
 [5]{.pnum} *Mandates:*
 
-  * [5.1]{.pnum} If `Extents​::​rank_dynamic() == 0` is `true`, then the size of the multidimensional index space `Extents()` is representable as a value of type typename `Extents​::​index_type`.
+  * [5.1]{.pnum} If `Extents​::​rank_dynamic() == 0` is `true`, then the size of the multidimensional index space `Extents()` is representable as a value of type `Extents​::​index_type`.
```
Comment (mhoemmen) [inline layout_padded/layout_padded.md:1651]: The other one says `index_type`, which I think is valid here.

```suggestion
  * [5.1]{.pnum} If `Extents​::​rank_dynamic() == 0` is `true`, then the size of the multidimensional index space `Extents()` is representable as a value of type `index_type`.
```

File: layout_padded/layout_padded.md (hunk 5 above)
Comment (mhoemmen) [reply layout_padded/layout_padded.md:2127]: I think `index_type` is valid here, btw.

File: layout_padded/layout_padded.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (nmm0) [reply layout_padded/layout_padded.md:2695]: yeah I agree with it, I know it's used to make the index easier to write but it's confusing

**Pull request merged**